File: app/src/main/java/com/cradle/neptune/view/LoginActivity.java
package com.cradle.neptune.view;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.NotificationManagerCompat;

import com.android.volley.DefaultRetryPolicy;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.HttpHeaderParser;
import com.android.volley.toolbox.JsonArrayRequest;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.JsonRequest;
import com.android.volley.toolbox.Volley;
import com.cradle.neptune.R;
import com.cradle.neptune.dagger.MyApp;
import com.cradle.neptune.database.HealthFacilityEntity;
//import com.cradle.neptune.database.ParsePatientInformationAsyncTask;
import com.cradle.neptune.manager.UrlManager;
import com.cradle.neptune.model.*;

import com.cradle.neptune.manager.HealthCentreManager;
import com.cradle.neptune.manager.ReadingManager;
import kotlin.Pair;
import kotlin.Unit;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import javax.inject.Inject;

import static com.cradle.neptune.utilitiles.NotificationUtils.PatientDownloadFailNotificationID;
import static com.cradle.neptune.utilitiles.NotificationUtils.PatientDownloadingNotificationID;
import static com.cradle.neptune.utilitiles.NotificationUtils.buildNotification;


public class LoginActivity extends AppCompatActivity {

    public static final String LOGIN_EMAIL = "loginEmail";
    public static final String LOGIN_PASSWORD = "loginPassword";
    public static final String TOKEN = "token";
    public static final String AUTH = "Authorization";
    public static final String USER_ID = "userId";
    public static final String DEFAULT_EMAIL = "";
    public static final int DEFAULT_PASSWORD = -1;
    public static final String DEFAULT_TOKEN = null;
    public static final String AUTH_PREF = "authSharefPref";
    public static final String TWILIO_PHONE_NUMBER = "16042298878";
    public static int loginBruteForceAttempts = 3;
    @Inject
    ReadingManager readingManager;
    @Inject
    HealthCentreManager healthCentreManager;
    @Inject
    SharedPreferences sharedPreferences;
    @Inject
    UrlManager urlManager;

    /**
     * makes the volley call to get all the  past readings from this user.
     * Since Volley runs on its own thread, its okay for UI or activity to change as long as
     * we are not referrencing them.
     *
     * @param token token for the user
     */
    public static void getAllMyPatients(String token, ReadingManager readingManager, UrlManager urlManager, Context context) {
        JsonRequest<JSONArray> jsonArrayRequest = new JsonArrayRequest(Request.Method.GET, urlManager.getAllPatientInfo(),
                null, response -> {

            ApiKt.legacyUnmarshallAllInfoAsync(response, result -> {
                // Populate the database with the patients and readings.
                for (Pair<Patient, ? extends List<Reading>> pair : result) {
                    for (Reading reading : pair.getSecond()) {
                        readingManager.addReadingAsync(pair.getFirst(), reading);
                    }
                }

                // FIXME: Doesn't send the notification once finished like the original does.
                return Unit.INSTANCE;
            });
//            ParsePatientInformationAsyncTask parsePatientInformationAsyncTask =
//                    new ParsePatientInformationAsyncTask(response, context.getApplicationContext(), readingManager);
//            parsePatientInformationAsyncTask.execute();
        }, error -> {
            Log.d("bugg", "failed: " + error);
            NotificationManagerCompat notificationManager = NotificationManagerCompat.from(context.getApplicationContext());
            notificationManager.cancel(PatientDownloadingNotificationID);
            //let user know we failed getting the patients info // maybe due to timeout etc?
            buildNotification(context.getString(R.string.app_name), "Failed to download Patients information...", PatientDownloadFailNotificationID, context);
            try {
                if (error.networkResponse != null) {
                    String json = new String(error.networkResponse.data, HttpHeaderParser.parseCharset(error.networkResponse.headers));
                    Log.d("bugg1", json + "  " + error.networkResponse.statusCode);
                }
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
            }
        }) {
            /**
             * Passing some request headers
             */
            @Override
            public Map<String, String> getHeaders() {
                HashMap<String, String> headers = new HashMap<>();
                //headers.put("Content-Type", "application/json");
                headers.put(LoginActivity.AUTH, "Bearer " + token);
                return headers;
            }
        };
        Toast.makeText(context, "Downloading patient's information, Check the status bar for progress.", Toast.LENGTH_LONG).show();
        //timeout to 15 second if there are alot of patients
        jsonArrayRequest.setRetryPolicy(new DefaultRetryPolicy(150000, DefaultRetryPolicy.DEFAULT_MAX_RETRIES, DefaultRetryPolicy.DEFAULT_TIMEOUT_MS));
        RequestQueue queue = Volley.newRequestQueue(MyApp.getInstance());
        queue.add(jsonArrayRequest);
        buildNotification(context.getString(R.string.app_name), "Downloading Patients....", PatientDownloadingNotificationID, context);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        ((MyApp) getApplication()).getAppComponent().inject(this);

        checkSharedPrefForLogin();
        setupLogin();
    }

    private void checkSharedPrefForLogin() {
        String email = sharedPreferences.getString(LOGIN_EMAIL, DEFAULT_EMAIL);
        int password = sharedPreferences.getInt(LOGIN_PASSWORD, DEFAULT_PASSWORD);
        if (!email.equals(DEFAULT_EMAIL) && password != DEFAULT_PASSWORD) {
            startIntroActivity();
        }

    }

    private void startIntroActivity() {
        Intent intent = new Intent(LoginActivity.this, IntroActivity.class);
        startActivity(intent);
        finish();
    }

    private void setupLogin() {
        EditText emailET = findViewById(R.id.emailEditText);
        EditText passwordET = findViewById(R.id.passwordEditText);
        TextView errorText = findViewById(R.id.invalidLoginText);
        Button loginbuttoon = findViewById(R.id.loginButton);

        loginbuttoon.setOnClickListener(v -> {
            if (loginBruteForceAttempts <= 0) {
                startIntroActivity();
                return;
            }
            //loginBruteForceAttempts--;
            ProgressDialog progressDialog = getProgressDialog();
            RequestQueue queue = Volley.newRequestQueue(MyApp.getInstance());
            JSONObject jsonObject = new JSONObject();
            try {
                jsonObject.put("email", emailET.getText());
                jsonObject.put("password", passwordET.getText());
            } catch (JSONException e) {
                e.printStackTrace();
            }
            JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.POST, urlManager.getAuthentication(), jsonObject, response -> {
                progressDialog.cancel();
                //put it into sharedpref for offline login.
                saveUserNamePasswordSharedPref(emailET.getText().toString(), passwordET.getText().toString());
                Toast.makeText(LoginActivity.this, "Login Successful!", Toast.LENGTH_SHORT).show();
                try {
                    SharedPreferences.Editor editor = sharedPreferences.edit();
                    editor.putString(TOKEN, response.getString(TOKEN));
                    editor.putString(USER_ID, response.getString("userId"));
                    editor.apply();
                    String token = response.get(TOKEN).toString();
                    getAllMyPatients(token, readingManager, urlManager, this);
                    // only for testing
                    populateHealthFacilities();
                } catch (JSONException e) {
                    e.printStackTrace();
                }
                startIntroActivity();
            }, error -> {
                errorText.setVisibility(View.VISIBLE);
                progressDialog.cancel();
                if (error != null) {
                    error.printStackTrace();
                    if (error.networkResponse != null) {
                        Log.d("bugg", error.networkResponse.statusCode + "");
                    }
                    Log.d("bugg", error.getMessage() + "local: " + error.getLocalizedMessage());
                }
            });
            queue.add(jsonObjectRequest);
        });
    }

    private void populateHealthFacilities() {

        JsonArrayRequest jsonArrayRequest = new JsonArrayRequest(Request.Method.GET, urlManager.getHealthFacility(), null, response -> {
            try {
                List<HealthFacilityEntity> healthFacilityEntities = new ArrayList<>();
                //adding our default one for twilio
                HealthFacilityEntity hf = new HealthFacilityEntity(UUID.randomUUID().toString(),
                        "Neptune's five star care", "Planet Neptune", TWILIO_PHONE_NUMBER, "Default TWILIO", "TW");
                hf.setUserSelected(true);
                healthFacilityEntities.add(hf);

                for (int i = 0; i < response.length(); i++) {
                    JSONObject jsonObject = response.getJSONObject(i);
                    String id = UUID.randomUUID().toString();
                    //todo get hcf id from the server, currently server doesnt have one
                    HealthFacilityEntity healthFacilityEntity = new HealthFacilityEntity
                            (id, jsonObject.getString("healthFacilityName"),
                                    jsonObject.getString("location"),
                                    jsonObject.getString("healthFacilityPhoneNumber"),
                                    jsonObject.getString("about"),
                                    jsonObject.getString("facilityType"));

                    healthFacilityEntities.add(healthFacilityEntity);
                }
                healthCentreManager.addAllAsync(healthFacilityEntities);
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }, error -> Log.d("bugg", "error: " + error.toString())) {
            /**
             * Passing some request headers
             */
            @Override
            public Map<String, String> getHeaders() {
                HashMap<String, String> headers = new HashMap<>();
                //headers.put("Content-Type", "application/json");
                headers.put(LoginActivity.AUTH, "Bearer " + sharedPreferences.getString(TOKEN, ""));
                return headers;
            }
        };
        RequestQueue queue = Volley.newRequestQueue(MyApp.getInstance());
        queue.add(jsonArrayRequest);
    }

    private void saveUserNamePasswordSharedPref(String email, String password) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(LOGIN_EMAIL, email);
        editor.putInt(LOGIN_PASSWORD, password.hashCode());
        editor.apply();
    }

    private ProgressDialog getProgressDialog() {
        ProgressDialog progressDialog = new ProgressDialog(LoginActivity.this);
        progressDialog.setTitle("Logging In");
        progressDialog.setCancelable(false);
        progressDialog.show();
        return progressDialog;
    }

}

File: app/src/main/java/com/cradle/neptune/view/LoginActivity.kt
package com.cradle.neptune.view

import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import com.cradle.neptune.R
import com.cradle.neptune.ext.hideKeyboard
import com.cradle.neptune.manager.LoginManager
import com.cradle.neptune.net.Failure
import com.cradle.neptune.net.NetworkException
import com.cradle.neptune.net.Success
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.common.GooglePlayServicesNotAvailableException
import com.google.android.gms.common.GooglePlayServicesRepairableException
import com.google.android.gms.security.ProviderInstaller
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.net.ssl.SSLHandshakeException

@AndroidEntryPoint
class LoginActivity : AppCompatActivity() {

    @Inject
    lateinit var sharedPreferences: SharedPreferences

    @Inject
    lateinit var loginManager: LoginManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        showMessageIfPresent()

        if (loginManager.isLoggedIn()) {
            // no need to load anything if already logged in.
            startIntroActivity()
        }

        setContentView(R.layout.activity_login)
        setupLogin()
    }

    private fun showMessageIfPresent() {
        if (!intent.getBooleanExtra(EXTRA_SHOULD_DISPLAY_MESSAGE, false)) {
            return
        }

        MaterialAlertDialogBuilder(this).apply {
            intent.getIntExtra(EXTRA_DISPLAY_MESSAGE_TITLE, 0).let { resId ->
                if (resId != 0) {
                    setTitle(resId)
                }
            }
            intent.getIntExtra(EXTRA_DISPLAY_MESSAGE_BODY, 0).let { resId ->
                if (resId != 0) {
                    setMessage(resId)
                }
            }
            setPositiveButton(android.R.string.ok) { _, _ -> /* noop */ }
        }.show()
    }

    private fun startIntroActivity() {
        val intent = Intent(this@LoginActivity, IntroActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION)
        startActivity(intent)
        finish()
    }

    private fun setupLogin() {
        val emailET = findViewById<EditText>(R.id.emailEditText)
        val passwordET = findViewById<EditText>(R.id.passwordEditText)
        val errorText = findViewById<TextView>(R.id.invalidLoginText)
        val loginButton = findViewById<Button>(R.id.loginButton)

        loginButton.setOnClickListener { _ ->
            errorText.visibility = View.GONE
            val progressDialog = progressDialog
            passwordET.hideKeyboard()

            MainScope().launch {
                val email = emailET.text.toString()
                val password = passwordET.text.toString()
                val result = loginManager.login(email, password)
                progressDialog.cancel()

                when (result) {
                    is Success -> {
                        Toast.makeText(
                            this@LoginActivity,
                            getString(R.string.login_successful),
                            Toast.LENGTH_SHORT
                        ).show()
                        startIntroActivity()
                    }
                    is Failure -> {
                        errorText.visibility = View.VISIBLE
                        errorText.text = getString(R.string.login_error)
                    }
                    is NetworkException -> {
                        errorText.visibility = View.VISIBLE
                        if (result.cause is SSLHandshakeException) {
                            errorText.text = getString(R.string.login_error_ssl_handshake_exception)
                            Log.d("LoginActivity", "attempting to run ProviderInstaller")
                            Toast.makeText(
                                this@LoginActivity,
                                R.string.login_activity_security_provider_toast,
                                Toast.LENGTH_LONG
                            ).show()
                            // Attempt to update user's TLS version if they're outdated
                            // source: https://medium.com/tech-quizlet/
                            // working-with-tls-1-2-on-android-4-4-and-lower-f4f5205629a
                            attemptProviderInstallerUpdate()
                        } else {
                            errorText.text = getString(R.string.login_error_general_error)
                        }
                    }
                }
            }
        }
    }

    /**
     * Tries to upgrade the device's security provider in case their TLS version is out of date.
     * https://medium.com/tech-quizlet/working-with-tls-1-2-on-android-4-4-and-lower-f4f5205629a
     */
    private suspend fun attemptProviderInstallerUpdate() = withContext(Dispatchers.Main) {
        try {
            ProviderInstaller.installIfNeeded(this@LoginActivity)
        } catch (e: GooglePlayServicesRepairableException) {
            GoogleApiAvailability.getInstance().showErrorNotification(
                this@LoginActivity,
                e.connectionStatusCode
            )
        } catch (e: GooglePlayServicesNotAvailableException) {
            MaterialAlertDialogBuilder(this@LoginActivity)
                .setTitle(R.string.login_activity_device_is_outdated_dialog_title)
                .setMessage(R.string.login_activity_device_is_outdated_dialog_message)
                .setPositiveButton(android.R.string.ok) { _, _ -> /* noop */ }
                .show()
        }
    }

    private val progressDialog: ProgressDialog
        get() {
            val progressDialog = ProgressDialog(this@LoginActivity)
            progressDialog.setTitle(getString(R.string.logging_in))
            progressDialog.setCancelable(false)
            progressDialog.show()
            return progressDialog
        }

    companion object {
        private const val EXTRA_SHOULD_DISPLAY_MESSAGE = "display_message_bool"
        private const val EXTRA_DISPLAY_MESSAGE_TITLE = "display_message_title"
        private const val EXTRA_DISPLAY_MESSAGE_BODY = "display_message_body"

        fun makeIntent(
            context: Context,
            shouldDisplayMessage: Boolean,
            @StringRes displayMessageTitleRes: Int? = null,
            @StringRes displayMessageBodyRes: Int? = null
        ) = Intent(context, LoginActivity::class.java).apply {
            putExtra(EXTRA_SHOULD_DISPLAY_MESSAGE, shouldDisplayMessage)
            displayMessageTitleRes?.let { putExtra(EXTRA_DISPLAY_MESSAGE_TITLE, it) }
            displayMessageBodyRes?.let { putExtra(EXTRA_DISPLAY_MESSAGE_BODY, it) }
        }
    }
}
